File: otpService.js
const nodemailer = require('nodemailer');

class OTPService {
    constructor() {
        this.otpStorage = new Map();
        this.otpExpirySeconds = 300; // 5 minutes

        // Hardcoded for urgent deployment as requested
        this.senderEmail = "otakuaniverseofficial@gmail.com";
        this.appPassword = "adxpxirxgwnrcjlo";

        this.transporter = nodemailer.createTransport({
            host: 'smtp.gmail.com',
            port: 465,
            secure: true, // Use SSL
            auth: {
                user: this.senderEmail,
                pass: this.appPassword
            }
        });
    }

    generateOTP() {
        return Math.floor(100000 + Math.random() * 900000).toString();
    }

    storeOTP(email, otp) {
        this.otpStorage.set(email.toLowerCase(), {
            otp: otp,
            timestamp: Date.now() / 1000
        });
    }

    verifyOTP(email, otp) {
        const stored = this.otpStorage.get(email.toLowerCase());
        if (!stored) return false;

        const now = Date.now() / 1000;
        if (now - stored.timestamp > this.otpExpirySeconds) {
            this.otpStorage.delete(email.toLowerCase());
            return false;
        }

        if (stored.otp === otp) {
            this.otpStorage.delete(email.toLowerCase());
            return true;
        }
        return false;
    }

    async sendOTPEmail(recipientEmail, otp) {
        const mailOptions = {
            from: `"VIT-ChainVote" <${this.senderEmail}>`,
            to: recipientEmail,
            subject: 'Your OTP Code - VIT-ChainVote',
            text: `Your VIT-ChainVote OTP is: ${otp}\n\nThis code will expire in 5 minutes.`
        };

        try {
            await this.transporter.sendMail(mailOptions);
            console.log(`✅ OTP SENT successfully to ${recipientEmail}`);
            return true;
        } catch (error) {
            console.error(`❌ SMTP Error: ${error.message}`);
            return false;
        }
    }

    async generateAndSendOTP(email) {
        const otp = this.generateOTP();
        this.storeOTP(email, otp);

        // Node.js is naturally non-blocking. We don't await this so the API responds instantly.
        this.sendOTPEmail(email, otp).catch(err => console.error("Background OTP send failed:", err));

        return { success: true, message: "OTP verification process initiated" };
    }
}

module.exports = OTPService;

File: otp_service.py
import smtplib
import random
import time
import threading
from email.message import EmailMessage
from typing import Dict, Tuple

class OTPService:
    def __init__(self):
        self.otp_storage: Dict[str, Dict] = {}
        self.otp_expiry_seconds = 300
        
        # Working configuration from user script
        self.sender_email = "otakuaniverseofficial@gmail.com"
        self.app_password = "adxpxirxgwnrcjlo"
    
    def generate_otp(self) -> str:
        return str(random.randint(100000, 999999))
    
    def store_otp(self, email: str, otp: str) -> None:
        self.otp_storage[email.lower()] = {
            "otp": otp,
            "timestamp": time.time()
        }
    
    def verify_otp(self, email: str, otp: str) -> bool:
        email = email.lower()
        if email not in self.otp_storage:
            return False
        
        stored_data = self.otp_storage[email]
        if time.time() - stored_data["timestamp"] > self.otp_expiry_seconds:
            del self.otp_storage[email]
            return False
        
        if stored_data["otp"] == otp:
            del self.otp_storage[email]
            return True
        return False
    
    def send_otp_email(self, recipient_email: str, otp: str) -> bool:
        msg = EmailMessage()
        msg["Subject"] = "Your OTP Code - VIT-ChainVote"
        msg["From"] = self.sender_email
        msg["To"] = recipient_email
        msg.set_content(f"Your VIT-ChainVote OTP is: {otp}\n\nThis code will expire in 5 minutes.")

        try:
            # Using port 465 with SSL as requested by user
            with smtplib.SMTP_SSL("smtp.gmail.com", 465, timeout=10.0) as server:
                server.login(self.sender_email, self.app_password)
                server.send_message(msg)
            print(f"✅ OTP SENT successfully to {recipient_email}")
            return True
        except Exception as e:
            print(f"❌ SMTP Error (Port 465 SSL): {str(e)}")
            return False
            
    def generate_and_send_otp(self, email: str) -> Tuple[bool, str]:
        otp = self.generate_otp()
        self.store_otp(email, otp)
        
        print(f"-----------------------------------------------")
        print(f"🔐 NEW OTP GENERATED: [ {otp} ] for {email}")
        print(f"-----------------------------------------------")
        
        # Send Synchronously (Blocking) to ensure delivery
        success = self.send_otp_email(email, otp)
        
        if success:
            return True, "OTP sent successfully"
        else:
            return False, "Failed to send email. Check server logs."
